### ISSUE TITLE
[BUG] `zero.Init` fails when some class is brought in dynamically

### ISSUE THREAD

**Issue by stas00** [microsoft/DeepSpeed#2812]
Once the issue of https://github.com/microsoft/DeepSpeed/issues/2811 is resolved we immediately get a new problem with dynamic class importation.

This time the situation can be first demonstrated with the real code of:

```
from transformers import AutoModel
with deepspeed.zero.Init(config_dict_or_path=ds_config):
     x = AutoModel.from_config(x)
```

though usually it's much more complex - I will show a more complete example below:

Here is a simple way to repro this use case with a simple dynamic class creation:

```
import torch
import deepspeed

ds_config = dict(train_batch_size=1, zero_optimization=dict(stage=3))

with deepspeed.zero.Init(config_dict_or_path=ds_config):
    # this is the same as `class MyModel(torch.nn.Module): pass` but it's done dynamically 
    x = type("MyModel", (torch.nn.Module,), {})
    model = x()

deepspeed_engine, *_ = deepspeed.initialize(model=model, config_params=ds_config)
```

```
$ deepspeed --num_gpus 1 test2.py

Traceback (most recent call last):
  File "test2.py", line 9, in <module>
    model = x()
  File "/mnt/nvme0/code/github/00optimize/DeepSpeed-optim-grad-accessors/deepspeed/runtime/zero/partition_parameters.py", line 414, in __exit__
    shutdown_init_context()
  File "/mnt/nvme0/code/github/00optimize/DeepSpeed-optim-grad-accessors/deepspeed/runtime/zero/partition_parameters.py", line 454, in shutdown_init_context
    _disable_class(subclass)
  File "/mnt/nvme0/code/github/00optimize/DeepSpeed-optim-grad-accessors/deepspeed/runtime/zero/partition_parameters.py", line 450, in _disable_class
    cls.__init__ = cls._old_init
AttributeError: type object 'MyModel' has no attribute '_old_init'

```

The problem here is that `zero.Init` context when entering overrides `__init__` methods, but because `MyModel` class wasn't loaded, `torch.nn.Module` doesn't have as its descendants and when `zero.Init` traverses the `torch.nn.Module`  descendants it can't install the override.

but on `exit` it tries to restore the init, yet it was never registered!

there are 2 problems here - the failure above, but of course even if I fix it by checking:

```
if hasattr(submodule,  '_old_init': cls.__init__ = cls._old_init
```
the problem is that weights of this class will be not be sharded during `zero.Init`.

So the only solution the user has here, is that they can't rely on `AutoModel`

They have to figure out what actual class they need to import it.

Just fixing the attribute issue in the traceback  would actually make things worse as it'd swipe the problem under the carpet. 

If the above condition `type object 'MyModel' has no attribute '_old_init'` occurs Deepspeed could assert and ask the user whether they were loading any modules dynamically inside the `zero.Init` context. And explain that all used `torch.nn.Module` subclasses must be imported into the space where `zero.Init` runs.

This could be very tricky for complex models that import dynamically inside their constructors. Again I'll use the example of  a [`VisionEncoderDecoderModel`](https://github.com/huggingface/transformers/blob/97d3390fc8edb210fcf0aad6a079406b018655b9/src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py#L191-L195) which is in essence like this:

```
VisionEncoderDecoderModel.from_pretrained(...)
```
which calls this constructor inside `zero.Init`
```
class VisionEncoderDecoderModel(...)::
    def __init__(self):
            encoder = AutoModel.from_config(config.encoder)
            decoder = AutoModelForCausalLM.from_config(config.decoder)
```

so here the user of `VisionEncoderDecoderModel` has no idea that a model internally calls on another model that will be imported dynamically at run-time after `zero.Init` context has been entered.

So this is going to break. and there are more and more models of this kind.

@tjruwase, @samyam

**Comment by tohtana** [comment]
Hi @stas00, I was considering a solution to both #2811 and this issue.

You wrote on #2811,

> Therefore wrt https://github.com/microsoft/DeepSpeed/issues/2812 I think a possible solution would be not to skip zero.Init's functionality if it's re-entrant, but to check which new nn.Module modules were added and if they don't have the DS hooks yet, insert them too.

But I was not sure when I should check if new ``nn.Module``s are added.  Users do not add a nested  ``zero.Init()`` for the dynamically defined class.

So how about using metaclass?
Here is an example code using metaclass.
The wrapper below just shows a message "wrapper called ..." but this should be replaced with partitioning of parameters in ``zero.Init()``.

```python
import torch
import functools


class WrappingMetaclass(type):
    def __new__(cls, clsname, bases, attrs):
        if all(not issubclass(base, torch.nn.Module) for base in bases) or "__init__" not in attrs:
            return super().__new__(cls, clsname, bases, attrs)

        attrs["old_init"] = attrs["__init__"]

        def wrap(f):
            @functools.wraps(f)
            def wrapper(module, *args, **kwargs):
                print(f"wrapper called self.__class__={module.__class__} {clsname}")
                f(module, *args, **kwargs)
            return wrapper

        attrs["__init__"] = wrap(attrs["old_init"])
        return super().__new__(cls, clsname, bases, attrs)


### Set metaclass
old_nn_module_cls = torch.nn.Module
old_type = type
torch.nn.Module = WrappingMetaclass("Module", (torch.nn.Module,), {})
type = WrappingMetaclass

### Pattern 1: declare a class
class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        return self.factor * x

model = Model()


### Pattern 2-1: Dynamically declare a class
DynamicModel = type("DynamicModel", (torch.nn.Module,), {"__init__": lambda self: super(DynamicModel, self).__init__()})
dy_model1 = DynamicModel()

### Pattern 2-2: Dynamically subclass an user-defined class
DynamicModelSub = type("DynamicModelSub", (Model,), {"__init__": lambda self: super(DynamicModelSub, self).__init__()})
dy_model2 = DynamicModelSub()

### Pattern 3: Dynamically import a class
"""
##In test_submodule.py:
class ExModule(torch.nn.Module):
    def __init__(self, factor):
        super(ExModule, self).__init__()

    def forward(self, x):
        return self.factor * x
"""
from test_submodule import ExModule
ex_model2 = ExModule()

torch.nn.Module = old_nn_module_cls
type = old_type

# New class is not wrapped any more
DynamicModel = type("DynamicModel", (torch.nn.Module,), {"__init__": lambda self: super(DynamicModel, self).__init__()})
dy_model3 = DynamicModel()
```

The outputs are:
```bash
wrapper called self.__class__=<class '__main__.Model'> Model
wrapper called self.__class__=<class '__main__.DynamicModel'> DynamicModel
wrapper called self.__class__=<class '__main__.DynamicModelSub'> DynamicModelSub 
wrapper called self.__class__=<class '__main__.DynamicModelSub'> Model # <= Wrapper of superclass's __init__
wrapper called self.__class__=<class 'test_submodule.ExModule'> ExModule
```

This approach does not seem very safe, but could be the only way to add a wrapper to dynamically defined classes in a non-intrusive manner. We can just ignore nested ``zero.Init()`` with this solution.

I would appreciate it if you could give me any suggestions.

**Comment by stas00** [comment]
> But I was not sure when I should check if new nn.Modules are added. Users do not add a nested zero.Init() for the dynamically defined class.

oh, but they do, re-pasting a section from https://github.com/microsoft/DeepSpeed/issues/2811



Here is the reduced version of their code that breaks:

```
from transformers import DonutProcessor, VisionEncoderDecoderModel
import torch
import deepspeed
from transformers.deepspeed import HfDeepSpeedConfig

ds_config = dict(train_batch_size=1, zero_optimization=dict(stage=3))

dschf = HfDeepSpeedConfig(ds_config)  # keep this object alive

model = VisionEncoderDecoderModel.from_pretrained("naver-clova-ix/donut-base-finetuned-cord-v2")

deepspeed_engine, *_ = deepspeed.initialize(model=model, config_params=ds_config)
```

You can see - there is nothing bad about this code.

When running this script - things go boom (infinite recursion), exactly as in the simple repro script at the beginning of this post.

and so what happens inside of it is this:

```
VisionEncoderDecoderModel.from_pretrained(...)
```
which calls this constructor inside `zero.Init`
```
class VisionEncoderDecoderModel(...)::
    def __init__(self):
            encoder = AutoModel.from_config(config.encoder)
            decoder = AutoModelForCausalLM.from_config(config.decoder)
```

where `from_config` calls `zero.Init` too! Nesting!

So users do use nested `zero.Init` as they aren't even aware `zero.Init` is called inside `from_config` and `from_pretrained`

> So how about using metaclass?

This looks like a great idea, @tohtana 

Does it work for the real example code a few paras up?

You also have to be very careful not to call partitioning too soon as this would break a lot of code. We changed the partitioning code to carefully wait in each class until all progeny finished their init. But I guess you will quickly discover such problems if you run the hf tests in the deepspeed CI.

----------

My idea was to make `zero.Init` re-entrant and simply install the special init for those new classes that have been discovered inside the nested `zero.Init` call. This is because when  `encoder` below is being created, by the time it gets to `zero.Init` the dynamicly imported classes have already been imported, so it's just a matter of carefully re-traversing the `nn.Module` subclasses but install the `init` override only for yet unseen sub-classes. 

```
class VisionEncoderDecoderModel(...)::
    def __init__(self):
            encoder = AutoModel.from_config(config.encoder)
            decoder = AutoModelForCausalLM.from_config(config.decoder)
```

But of course if your idea works, this is just as good. I'm not attached to either way.

**Comment by stas00** [comment]
I think we have an unrelated problem which probably requires a separate issue. And that's when a user wants to work with 2 independent models. I have no idea how deepspeed would handle that. Create 2 engines? I'm not sure that even works, as if I remember there are some globals and some assumptions that there is only one engine.

Why am I bringing this up here? Because `zero.Init` happens before `deepspeed.initialize` and if 2 models are unrelated ideally each `zero.Init` should be an independent context and not cross-pollute partitioning a model it's not going to work with. Not sure. I'm just bringing this up while we decide how to make `zero.Init` more flexible - should this issue be tackled down the road.

**Comment by tohtana** [comment]
@stas00 Thank you for your detailed explanation. Let me clarify some points.

When using `VisionEncoderDecoderModel`, you mean the structure is:
```
with zero.Init():
  # (Dynamically) define a new class
  with zero.Init():
    # Instantiate the new class
```
Is my understanding correct? 
If this is all the requirements we should meet, I think we should take your approach. It is much simpler and easier.

My concern is that we would need to handle the following case in the near future:
```
with zero.Init():
  # (Dynamically) define a new class
  # Instantiate the new class
```
In this pattern, we don't have a chance to add a wrapper to the new class.
What do you think about this?

And, regarding multiple models, I agree with you. I believe we need to work on it separately.

@jeffra @tjruwase @samyam 
I don't know about expected behaviors of the deepspeed engine when it is given multiple models.
Can you give us some inputs?

**Comment by stas00** [comment]
> Is my understanding correct?

It is this:
```
# dynamically import main class A
with zero.Init():
    # Instantiate  class A

    # dynamically import class B1
    with zero.Init():
        # Instantiate class B1 as a member of A

    # dynamically iimport  class B2
    with zero.Init():
        # Instantiate class B2 as a member of A
```

I'm sure there could be further nested (more than 1 level), but I'm yet to see such a model in practice. But let's solve this as a generic problem.

> My concern is that we would need to handle the following case in the near future:
```
with zero.Init():
  # (Dynamically) define a new class
  # Instantiate the new class
```

Specifically in the case of HF `transformers` standard models this is not the case. But I can absolutely see how this could happen in another user's model.

So definitely let's go with the most flexible solution. Though `deepspeed` could impose a requirement that all classes should be visible before `zero.Init` call and if this is not the case, it shouldn't fail to run, but either just not partition those classes until the first `forward` or assert and tell the user that a class was discovered that wasn't there during `zero.Init` and please do this or that. I'd favour the latter approach.

**Comment by tohtana** [comment]
Hi @stas00, I discussed within the DeepSpeed team how to fix the issues you reported.

As the result, for the nested `zero.Init()`, we will implement the approach you proposed as the first step. 
We will add wrappers to classes that are dynamically defined inside `zero.Init()` as you proposed. I think we also need to describe in the document that users are required to define a module class before `zero.Init()` context.

On the other hand, we still think we need to handle the following usage. 
```python
with zero.Init():
  # (Dynamically) define a new class
  # Instantiate the new class
```
My proposal using metaclass can handle this but takes a longer time to implement and test. So we will implement my approach as the next step.

Regarding use of multiple models and engines, I confirmed with @tjruwase that the following usage you mentioned is not officially supported.
```
dynamically import main class A
with zero.Init():
    a = Instantiate class A

    dynamically import class B1 via a.config.b1_name
    with zero.Init():
        a.b1 = Instantiate class B1 as a member of A

    dynamically import  class B2 via a.config.b2_name
    with zero.Init():
        a.b2 = Instantiate class B2 as a member of A
```

I will briefly check the behavior of this pattern and then file it as a new issue.

**Comment by stas00** [comment]
Sounds good, @tohtana  - I'm just not sure I follow your last comment as it contradicts the first comment. The first and the last are the same, no?

And if the last isn't supported many transformers models will not run under Deepspeed Z3 then without an advanced user intervention. This is of course your call whether you want to support this or not. There are more and more models emerging of this kind.

**Comment by tjruwase** [comment]
@stas00, I chatted extensively with @tohtana, and perhaps I can provide some clarification here. I think `not officially supported`' means `not anticipated` and `not tested`. We are both unsure of this pattern is handled, and @tohtana plans to investigate so we can learn more. Issues will be filed as appropriate. It sounds like this pattern is widely utilized by transformer models, which probably suggests all is well. But we just want to make sure. This investigation could probably result in appropriate unit tests.

**Comment by tohtana** [comment]
@stas, @tjruwase 
Sorry for the confusion. Let me also add some explanations.

The first part of my comment was about nesting `zero.Init()` and dynamically defined class inside `zero.Init()`.
The code I mentioned in the last part contained substitutions to attributes of an existing instance. This looked different from my viewpoint.

**Comment by stas00** [comment]
@tohtana, they aren't substitutes, they are the first time definition. A substitute would indeed be a huge issue (which in fact we need to talk about at some point too, since some models have - completely unrelated to the issue at hand)

I again invite you to look at the concrete example here:

https://github.com/huggingface/transformers/blob/97d3390fc8edb210fcf0aad6a079406b018655b9/src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py#L191-L195

it's very simple - I'm not asking to read the rest of the model, just those 4 lines.

And we can also test your experiments on the real cases reported by users like this one: https://github.com/huggingface/transformers/issues/21538

**Comment by stas00** [comment]
Thank you for clarifying, @tjruwase. We already know these patterns aren't supported, hence this Issue. All I was trying to communicate is that it's up to you what you consider supportable and what not. Apologies if my comment sounded like a challenge, I only meant to say that I wasn't insisting in case it was something too much to ask for.

I think this types of models is going to be quite wide-spread since multi-modal domain is growing very fast. e.g. the project I work on right now combines one LM and one CV models with a glue code in between. So I had to do a lot of acrobatics to make it work with ZeRO. and I have full control over the source code. I can show you the code if you're interested.

**Comment by tohtana** [comment]
@stas00 Thank you for the clarification. I understand that the code defines new attributes. Sorry, I just didn't properly understand the meaning of `substitute` and used the word to mean setting a value to somewhere.

Since it is difficult to estimate the cost for implementation of my approach using metaclass, we still want to implement your approach now. I think it is clear that your proposal can solve the issues of nested `zero.Init` and dynamically defined class.

As @tjruwase mentioned, I will also check if the case like vision transformer you mentioned works.

**Comment by stas00** [comment]
That sounds like a wise plan of action to me, @tohtana - thank you for working on this. You will make many users very happy!

**Comment by tohtana** [comment]
Hello @stas00,

I just submitted a PR (#2989) to fix the problems addressed in this issue (and #2812).

I checked the fix with some patterns.
Nested `zero.Init()`:
```
with deepspeed.zero.Init(config_dict_or_path=ds_config):
    with deepspeed.zero.Init(config_dict_or_path=ds_config):
        m1 = torch.nn.Linear(1,1)

deepspeed_engine, *_ = deepspeed.initialize(model=m1, config_params=ds_config)
```

Dynamically defined class:
```
with deepspeed.zero.Init(config_dict_or_path=ds_config):
    
    class MyModel(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.fc = torch.nn.Linear(1,1)

    with deepspeed.zero.Init(config_dict_or_path=ds_config):
        model = MyModel()

deepspeed_engine, *_ = deepspeed.initialize(model=model, config_params=ds_config)
```

The code in [the issue](https://github.com/huggingface/transformers/issues/21538) regarding `VisionEncoderDecoderModel` also ran after a few small fixes (it didn't converge in the settings).

As we discussed, the following pattern is not supported. We need another zero.Init() before the instantiation of the new class.
```
with deepspeed.zero.Init(config_dict_or_path=ds_config):
    class MyModel(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.fc = torch.nn.Linear(1,1)
    model = MyModel()

deepspeed_engine, *_ = deepspeed.initialize(model=model, config_params=ds_config)
```

In this pattern, `zero.Init()` finds that the new class wasn't properly processed and throws an error.
```
RuntimeError: `<class '__main__.MyModel'>' was not properly set up for sharding by zero.Init(). A subclass of torch.nn.Module must be defined before zero.Init() where an instance of the class is created.
```

I appreciate your support for our project. Any feedback is welcome.

**Comment by stas00** [comment]
I was only able to test a few cases, I will continue testing/studying your work tomorrow, @tohtana - thank you for working on it.

**Comment by tohtana** [comment]
Hi @stas00  , can we merge #2989? Do you still want to test?

**Comment by stas00** [comment]
Apologies that tomorrow turned into 2 weeks, @tohtana - I was trying to find a big enough slot to give it a serious attention and just wasn't manifesting.

I was also hoping that the 2 users who reported the problem would validate that your PR fixed their problem. But alas that didn't happen.

Attending to it now. Will post an update.

**Comment by tjruwase** [comment]
Fixed by #2989. Will open new issue for #3202 as needed.
